### PR TITLE
add condition and event info for not upgradable pods when update sidecarset (#1272)

### LINKED ISSUE #1272: [BUG] Update SidecarSet(Not just the image field) will not trigger sidecar version update.
**What happened**:
We have a SidecarSet deployed in Minikube.
- Changed SidecarSet.spec.containers[x].resources.limits.cpu.
- Upgrade the SidecarSet.spec.containers[x].image to a new version.

**What you expected to happen**:
The sidecar container version should be updated to the new version without restarting the pod.

**How to reproduce it (as minimally and precisely as possible)**:

- Install openKruise in minikube: https://openkruise.io/docs/installation/
- kubectl apply -f sidecarset.yaml

```
sidecarset.yaml
apiVersion: apps.kruise.io/v1alpha1
kind: SidecarSet
metadata:
  name: test-sidecarset3
spec:
  selector:
    matchLabels:
      sidecar-envoy: envoy
  updateStrategy:
    type: RollingUpdate
    maxUnavailable: 1
  containers:
  - name: sidecar12
    image: centos:6.7
    resources:
      requests:
        memory: "100Mi"
        cpu: "250m"
      limits:
        memory: "128Mi"
        cpu: "1000m"
    command: ["sleep", "998d"] # do nothing at all
    volumeMounts:
    - name: log-volume
      mountPath: /var/log
  volumes: # this field will be merged into pod.spec.volumes
  - name: log-volume
    emptyDir: {}
```
- kubectl apply -f cloneset.yaml
```
apiVersion: apps.kruise.io/v1alpha1
kind: CloneSet
metadata:
  labels:
    app: sample
  name: sample
spec:
  replicas: 2
  selector:
    matchLabels:
      app: sample
  template:
    metadata:
      labels:
        app: sample
        label1: v1
        label2: v2
    spec:
      containers:
      - name: nginx
        image: nginx:alpine
        resources:
          limits:
            cpu: 251m
            memory: 100Mi
          requests:
            cpu: 250m
            memory: 100Mi
```
- Changed SidecarSet.spec.containers[x].resources.limits.cpu.
- Upgrade the SidecarSet.spec.containers[x].image to a new version.

**Anything else we need to know?**:
I did found the "kruise-daemon" is failed.
```
kubectl get pod -n kruise-system
NAME                                         READY   STATUS             RESTARTS         AGE
kruise-controller-manager-5bcc8987fd-bskn6   1/1     Running            1 (7d7h ago)     8d
kruise-controller-manager-5bcc8987fd-tttdr   1/1     Running            2 (4d9h ago)     8d
kruise-daemon-w2pwr                          0/1     CrashLoopBackOff   1204 (98s ago)   8d
```

**Environment**:
- Kruise version: 1.4.0
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"25", GitVersion:"v1.25.4", GitCommit:"872a965c6c6526caa949f0c6ac028ef7aff3fb78", GitTreeState:"clean", BuildDate:"2022-11-09T13:36:36Z", GoVersion:"go1.19.3", Compiler:"gc", Platform:"darwin/amd64"}
Kustomize Version: v4.5.7
Server Version: version.Info{Major:"1", Minor:"26", GitVersion:"v1.26.1", GitCommit:"8f94681cd294aa8cfd3407b8191f6c70214973a4", GitTreeState:"clean", BuildDate:"2023-01-18T15:51:25Z", GoVersion:"go1.19.5", Compiler:"gc", Platform:"linux/amd64"}
```
- Install details (e.g. helm install args): follow https://openkruise.io/docs/installation/
- Others:

### PR DESCRIPTION
**Pull request by MarkLux**
### Ⅰ. Describe what this PR does

Currently the sidecar container in-place upgrade can only support changing of `image` field, if user changed other fields, the injected pods will not execute sidecar container upgrade. 

However, this rule is not explicit for users, they can not get any hint or info in this situation. This PR add an event for pods, when the injected sidecar  changed but cannot be upgrade. If user changed SidecarSet but the pods are not upgraded, they can use `kubectl describe sidecarset <sidecarset-name>` or `kubectl describe pod <pod-name>` to see the event and figure out the reason.

### Ⅱ. Does this pull request fix one issue?

fixes #1272 

### Ⅲ. Describe how to verify it

1. create a sidecar set

```
apiVersion: apps.kruise.io/v1alpha1
kind: SidecarSet
metadata:
  name: demo-sidecarset
spec:
  selector:
    matchLabels:
      app: demo-server
  updateStrategy:
    type: RollingUpdate
  containers:
  - name: sidecar
    image: nginx
```

2. create some pods with label (here I using a CloneSet), the pods will be injected with sidecar container immediately

```
apiVersion: apps.kruise.io/v1alpha1
kind: CloneSet
metadata:
  labels:
    app: demo-server
  name: demo-server
spec:
  updateStrategy:
    type: InPlaceIfPossible
  replicas: 3
  selector:
    matchLabels:
      app: demo-server
  template:
    metadata:
      labels:
        app: demo-server
    spec:
      containers:
      - name: demo-server
        image: tomcat
        ports:
        - containerPort: 8080
```

3. update the SidecarSet , change some fields more than `image` (here I change the `image` and add a `command` field)

```
apiVersion: apps.kruise.io/v1alpha1
kind: SidecarSet
metadata:
  name: demo-sidecarset
spec:
  selector:
    matchLabels:
      app: demo-server
  updateStrategy:
    type: RollingUpdate
  containers:
  - name: sidecar
    image: busybox
    command: [ "/bin/sh", "-c", "touch a.txt && tail -f a.txt"]
```

4. apply  the changed sidecarset, the pods will not be upgraded, using `kubectl  get pod `, we can see the condition info:

```
- lastProbeTime: null
    lastTransitionTime: "2023-06-19T05:41:13Z"
    message: Pod's sidecar set is not upgradable due to changes out of image field
    reason: UpdateImmutableField
    status: "False"
    type: SidecarSetUpgradable
```

5. and the SidecarSet got an event for not upgradable pods

```
Events:
  Type    Reason             Age                From                   Message
  ----    ------             ----               ----                   -------
  Normal  NotUpgradablePods  5s (x5 over 161m)  sidecarset-controller  SidecarSet in-place update detected 2 not upgradable pod(s) in this round, will skip them.
```

### CODE REVIEW

**Comment by veophi** [comment]
@MarkLux  How about a condition for such Pod instead of an event? (too many events will oppress ETCD）

**Comment by MarkLux** [comment]
> @MarkLux How about a condition for such Pod instead of an event? (too many events will oppress ETCD）

like this?

```
Conditions:
  Type                 Status
  KruisePodReady       True
  InPlaceUpdateReady   True
  Initialized          True
  Ready                True
  ContainersReady      True
  PodScheduled         True
  SidecarUpgradble       False
```

But the Condition Field may be too short to show the reason, and it's not obvious. I think after user updated SidecarSet, they need a feedback.

If too many events can make an extra stress to etcd, how about add event for SidecarSet instead of Pods? (thus we only got an single event for each update)

Or maybe we can give a combination of event and condition: add  condition for pods & add event for sidecar?

**Comment by veophi** [comment]
> If too many events can make an extra stress to etcd, how about add event for SidecarSet instead of Pods? (thus we only got an single event for each update)
> 
> btw, if the sidecar upgrade succeeded, an event will be added to all the pods(by kubectl), I wonder if these events can will oppress etcd too?

@MarkLux  `Successful Update` event will be sent only when Pod update successfully. But `Failed Update` events may be added at Each Reconciling.

```yaml
conditions:
- type: SidecarSetUpgradable
   status: False
   reason: UpdateImmutableField
   message: Pod env and volume fields cannot be modified
```

**Comment by zmberg** [comment]
@MarkLux I have different idea, If SidecarSet manage more than one thousand pods, kruise will patch all the pods. Can we add the event and condition in SidecarSet?

File: pkg/controller/sidecarset/sidecarset_strategy_test.go [hunk 1]
```diff
@@ -524,7 +524,7 @@ func testSortNextUpgradePods(t *testing.T, factoryPods FactoryPods, factorySidec
 		t.Run(cs.name, func(t *testing.T) {
 			control := sidecarcontrol.New(cs.getSidecarset())
 			pods := cs.getPods()
-			injectedPods := strategy.GetNextUpgradePods(control, pods)
+			injectedPods, _ := strategy.GetNextUpgradePods(control, pods)
```
Comment (zmberg) [inline pkg/controller/sidecarset/sidecarset_strategy_test.go:527]: Please add UT for notUpgradablePods

**Comment by MarkLux** [comment]
@zmberg @veophi the code is updated, could you give a review?

File: pkg/controller/sidecarset/sidecarset_processor.go [hunk 2]
```diff
@@ -197,8 +220,8 @@ func (p *Processor) updatePodSidecarAndHash(control sidecarcontrol.SidecarContro
 			klog.Errorf("sidecarSet(%s) patch pod(%s/%s) metadata failed: %s", sidecarSet.Name, podClone.Namespace, podClone.Name, err.Error())
 			return err
 		}
-		//update pod in store
-		return p.Client.Update(context.TODO(), podClone)
+		// update pod in store, both status and spec
+		return p.Client.Status().Update(context.TODO(), podClone)
 	})
 	return err
```
Comment (zmberg) [inline pkg/controller/sidecarset/sidecarset_processor.go:226]:
```
        if err != nil {
		return err
	}
	// patch SidecarSetUpgradable condition
	if isEqual := podutil.UpdatePodCondition(&podClone.Status, &corev1.PodCondition{
		Type:   SidecarSetUpgradable,
		Status: corev1.ConditionTrue,
	}); isEqual {
		return nil
	}
	_, condition := podutil.GetPodCondition(&podClone.Status, SidecarSetUpgradable)
	body := fmt.Sprintf(`{"status":%s}`, util.DumpJSON(condition))
	return p.Client.Status().Patch(context.TODO(), podClone, client.RawPatch(types.MergePatchType, []byte(body)))
```

**Comment by zmberg** [comment]
@MarkLux You can check the pod condition in [e2e code](https://github.com/openkruise/kruise/blob/master/test/e2e/apps/sidecarset.go#L741).

**Comment by veophi** [comment]
@MarkLux  If there are more than one SidecarSets that all select one Pod, you must refresh this condition of the Pod to `true` only if all the SidecarSets can upgrade the Pod. 

You can refer to [this] to fix your codes (https://github.com/openkruise/kruise/blob/ffe60f2600947bd8208c79400487f5b4d3b150b6/pkg/util/podreadiness/pod_readiness_utils.go#L98

**Comment by MarkLux** [comment]
> @MarkLux If there are more than one SidecarSets that all select one Pod, you must refresh this condition of the Pod to `true` **only if all the SidecarSets can upgrade the Pod**.
> 
> You can refer to [the following codes] to fix your logic (
> 
> https://github.com/openkruise/kruise/blob/ffe60f2600947bd8208c79400487f5b4d3b150b6/pkg/util/podreadiness/pod_readiness_utils.go#L98

got it

**Comment by MarkLux** [comment]
@zmberg PTAL

File: pkg/controller/sidecarset/sidecarset_processor.go [hunk 3]
```diff
@@ -602,3 +621,104 @@ func inconsistentStatus(sidecarSet *appsv1alpha1.SidecarSet, status *appsv1alpha
...
+	}
+
+	// patch SidecarSetUpgradable condition
+	if conditionChanged := podutil.UpdatePodCondition(&upgradablePod.Status, condition); !conditionChanged {
+		// reduce unnecessary patch.
+		return nil
+	}
+
+	mergePatch := fmt.Sprintf(`{"status": {"conditions": [%s]}}`, util.DumpJSON(condition))
+	return p.Client.Status().Patch(context.TODO(), upgradablePod, client.RawPatch(types.StrategicMergePatchType, []byte(mergePatch)))
```
Comment (zmberg) [inline pkg/controller/sidecarset/sidecarset_processor.go:676]: If patch success, please add log

File: pkg/controller/sidecarset/sidecarset_processor.go [hunk 4]
```diff
@@ -602,3 +621,104 @@ func inconsistentStatus(sidecarSet *appsv1alpha1.SidecarSet, status *appsv1alpha
...
+		if err != nil {
+			return err
+		}
+
+		// update pod condition
+		conditionUpdateResult := podutil.UpdatePodCondition(&podClone.Status, condition)
+		if !conditionUpdateResult {
+			return nil
+		}
+		err = p.Client.Status().Update(context.TODO(), podClone)
```
Comment (zmberg) [inline pkg/controller/sidecarset/sidecarset_processor.go:715]: Why not use patch function?

**Pull request merged**